gr-filter/lib/mmse_resampler_impl.cc

```diff
@@ -0,0 +1,142 @@
+ *
+ */
+
+#include "mmse_resampler_impl.h"
+#include <gnuradio/filter/mmse_resampler.h>
+#include <gnuradio/gr_complex.h>
+#include <gnuradio/io_signature.h>
+#include <gnuradio/thread/thread.h>
+#include <stdexcept>
+
+namespace gr {
+namespace filter {
+
+template <typename sample_t>
+typename mmse_resampler<sample_t>::sptr mmse_resampler<sample_t>::make(float phase_shift,
+                                                                       float resamp_ratio)
+{
+    return gnuradio::make_block_sptr<mmse_resampler_impl<sample_t>>(phase_shift,
+                                                                    resamp_ratio);
+}
+
+template <typename sample_t>
+mmse_resampler_impl<sample_t>::mmse_resampler_impl(float phase_shift, float resamp_ratio)
+    : block("mmse_resampler",
+            io_signature::make2(1, 2, sizeof(sample_t), sizeof(float)),
+            io_signature::make(1, 1, sizeof(sample_t))),
+      d_mu(phase_shift),
+      d_delta_mu(resamp_ratio)
+{
+    if (resamp_ratio <= 0)
+        throw std::out_of_range("resampling ratio must be > 0");
+    if (phase_shift < 0 || phase_shift > 1)
+        throw std::out_of_range("phase shift ratio must be > 0 and < 1");
+
+    this->set_inverse_relative_rate(d_delta_mu);
+    this->message_port_register_in(pmt::intern("msg_in"));
+    this->set_msg_handler(pmt::intern("msg_in"),
+                          [this](pmt::pmt_t msg) { this->handle_msg(msg); });
+}
+
+template <typename sample_t>
+void mmse_resampler_impl<sample_t>::handle_msg(const pmt::pmt_t& msg)
+{
+    if (!pmt::is_dict(msg))
+        return;
+    // set resamp_ratio or mu by message dict
+    if (pmt::dict_has_key(msg, pmt::intern("resamp_ratio"))) {
+        set_resamp_ratio(pmt::to_float(pmt::dict_ref(
+            msg, pmt::intern("resamp_ratio"), pmt::from_float(resamp_ratio()))));
+    }
+    if (pmt::dict_has_key(msg, pmt::intern("mu"))) {
+        set_mu(
+            pmt::to_float(pmt::dict_ref(msg, pmt::intern("mu"), pmt::from_float(mu()))));
+    }
+}
+
+template <typename sample_t>
+void mmse_resampler_impl<sample_t>::forecast(int noutput_items,
+                                             gr_vector_int& ninput_items_required)
+{
+    unsigned ninputs = ninput_items_required.size();
+    for (unsigned i = 0; i < ninputs; i++) {
+        ninput_items_required[i] =
+            (int)ceil((noutput_items * d_delta_mu) + d_resamp.ntaps()) + d_delta_idx;
+    }
+}
+
+template <typename sample_t>
+int mmse_resampler_impl<sample_t>::general_work(int noutput_items,
+                                                gr_vector_int& ninput_items,
+                                                gr_vector_const_void_star& input_items,
+                                                gr_vector_void_star& output_items)
+{
+    const sample_t* in = (const sample_t*)input_items[0];
+    sample_t* out = (sample_t*)output_items[0];
+    const auto max_input_index = ninput_items[0] - d_resamp.ntaps();
```

Comment:
For the two-input case, is there something in the scheduler that ensures that `ninput_items[0] == ninput_items[1]`? If not, then probably this should be `std::min(ninput_items[0], ninput_items[1])` in this case.

Reply:
oooh yes, indeed.